Don't assume image will appear on screen for pixmap_scale

abewieland opened the pull request:
Fixes #244.

artemsen commented on src/pixmap_scale.c:
```diff
@@ -559,6 +559,12 @@ void pixmap_scale(enum pixmap_aa_mode scaler, const struct pixmap* src,
                   struct pixmap* dst, ssize_t x, ssize_t y, float scale,
                   bool alpha)
 {
+    // Do nothing if the image won't appear on the screen
```
Unfortunately this fix doesn't resolve the issue, the application still crashes.
This is because not all, but only part of the image may be outside the window.

Here the log with enabled address sanitizer:
```
==3435==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x75a334f20b80 at pc 0x6065137b51e8 bp 0x7ffe53750f10 sp 0x7ffe53750f08
READ of size 4 at 0x75a334f20b80 thread T0
    #0 0x6065137b51e7 in apply_hk swayimg.244/.build/../src/pixmap_scale.c:308:21
    #1 0x6065137b3b17 in sc_task swayimg.244/.build/../src/pixmap_scale.c:426:5
    #2 0x6065137b2f6c in pixmap_scale_aa swayimg.244/.build/../src/pixmap_scale.c:546:5
    #3 0x6065137b1e2f in pixmap_scale swayimg.244/.build/../src/pixmap_scale.c:590:9
    #4 0x6065137bfff5 in draw_image swayimg.244/.build/../src/viewer.c:617:9
    #5 0x6065137bd4a0 in redraw swayimg.244/.build/../src/viewer.c:649:9
    #6 0x6065137bcec0 in viewer_handle swayimg.244/.build/../src/viewer.c:845:13
    #7 0x606513798b60 in handle_event_queue swayimg.244/.build/../src/application.c:132:13
    #8 0x606513799182 in app_run swayimg.244/.build/../src/application.c:529:17
    #9 0x6065137ace1c in main swayimg.244/.build/../src/main.c:175:14
    #10 0x75a3454fde07 in __libc_start_call_main /usr/src/debug/glibc/glibc/csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #11 0x75a3454fdecb in __libc_start_main /usr/src/debug/glibc/glibc/csu/../csu/libc-start.c:360:3
    #12 0x606513657f44 in _start (swayimg.244/.build/swayimg+0x48f44) (BuildId: 2b7ebe089796bdd11f38bcdeaa5eb2336f2c2f47)

0x75a334f20b80 is located 3200 bytes before 1920000-byte region [0x75a334f21800,0x75a3350f6400)
allocated by thread T0 here:
    #0 0x60651374ae19 in calloc (swayimg.244/.build/swayimg+0x13be19) (BuildId: 2b7ebe089796bdd11f38bcdeaa5eb2336f2c2f47)
    #1 0x6065137aeeca in pixmap_create swayimg.244/.build/../src/pixmap.c:15:20
    #2 0x6065137a420d in image_allocate_frame swayimg.244/.build/../src/image.c:145:9
    #3 0x6065137d4316 in decode_single swayimg.244/.build/../src/formats/png.c:83:25
    #4 0x6065137d3bbb in decode_png swayimg.244/.build/../src/formats/png.c:328:14
    #5 0x6065137acd32 in image_from_memory swayimg.244/.build/../src/loader.c:186:18
    #6 0x6065137ac56d in image_from_file swayimg.244/.build/../src/loader.c:228:14
    #7 0x6065137abf14 in loader_from_source swayimg.244/.build/../src/loader.c:337:18
    #8 0x6065137ac6a7 in loader_from_index swayimg.244/.build/../src/loader.c:356:18
    #9 0x60651379857c in load_first_file swayimg.244/.build/../src/application.c:243:18
    #10 0x606513797a72 in app_init swayimg.244/.build/../src/application.c:394:19
    #11 0x6065137ace03 in main swayimg.244/.build/../src/main.c:171:10
    #12 0x75a3454fde07 in __libc_start_call_main /usr/src/debug/glibc/glibc/csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #13 0x75a3454fdecb in __libc_start_main /usr/src/debug/glibc/glibc/csu/../csu/libc-start.c:360:3
    #14 0x606513657f44 in _start (swayimg.244/.build/swayimg+0x48f44) (BuildId: 2b7ebe089796bdd11f38bcdeaa5eb2336f2c2f47)

SUMMARY: AddressSanitizer: heap-buffer-overflow swayimg.244/.build/../src/pixmap_scale.c:308:21 in apply_hk
Shadow bytes around the buggy address:
  0x75a334f20900: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x75a334f20980: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x75a334f20a00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x75a334f20a80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x75a334f20b00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x75a334f20b80:[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x75a334f20c00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x75a334f20c80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x75a334f20d00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x75a334f20d80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x75a334f20e00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==3435==ABORTING
```

abewieland replied:
The code correctly handles images that are partially onscreen (that's also why I added some tests). I had just made a rounding error in the check - the new force-pushed commit should fix things.

artemsen commented:
Unfortunately no =) The latest commit also doesn't resolve the issue. It only handles one case - when the entire image is offscreen.
But when at least one line is visible - program crashes. You need to add to apply_vk/apply_hk something like what I did in [pixmap_copy](https://github.com/artemsen/swayimg/blob/e1509c327d12531b54ae94ce8b30be3ea939b33c/src/pixmap.c#L194) - constrain the rectangle to process.

PS If suddenly you don't have time or desire to fix this - let me know, I'll do it a little later.

abewieland commented:
Do you have an example of how I can recreate such a crash? apply_vk and apply_hk shouldn't have to deal with constraints on the input or output, because those are already done in new_kernel (where min and max are used to ensure the inputs and outputs remain within the relative pixmaps).

artemsen commented:
Steps are described in #244, but it's hard to reproduce on release build and small images.
Please, build debug version with address sanitizer (I use clang):
```
export CC="clang"
export CFLAGS="-g -fsanitize=address"
export LDFLAGS="-fsanitize=address"
meson --buildtype=debug ...
```

Example image I use for test:
![photo](https://github.com/user-attachments/assets/0c26156c-3f25-4d2d-9529-2b34ccce282a)

abewieland commented:
I still can't reproduce - for reference I've tried with both gcc and clang, following exactly this sequence (to ensure I know exactly what code I'm working with):
```
git clone https://github.com/abewieland/swayimg -b offscreen clean_clone
cp img.jpg clean_clone
cd clean_clone

export CC="gcc"
export CFLAGS="-g -fsanitize=address"
export LDFLAGS="-fsanitize=address"
meson setup --buildtype=debug build
meson compile -C build
./build/swayimg -c viewer.fixed=no img.jpg

rm -r build
export CC="clang"
meson setup --buildtype=debug build
meson compile -C build
./build/swayimg -c viewer.fixed=no img.jpg
``` 
I tried to move and resize in both cases for about 2 minutes straight and couldn't produce a segfault.

artemsen commented:
Sorry, the last fix actually works. I accidentally launched the wrong instance of the program. Thanks!

abewieland commented:
No problem. I'm just glad it wasn't some bug dependent on compiler version or system or something like that.
